virttest/virsh.py

```diff
@@ -621,6 +624,93 @@ def kosher_args(remote_ip, uri):
         return True not in all_false
 
 
+class EventNotFoundError(Exception):
+    """
+    Error when certain event cannot be found.
+    """
+
+    def __init__(self, details=''):
+        self.details = details
+        Exception.__init__(self)
+
+    def __str__(self):
+        return str(self.details)
+
+
+class EventTracker(object):
+
+    @staticmethod
+    def start_get_event(vm_name):
+        """
+        Use a virsh session with subcommand 'event' to catch events
+
+        :param vm_name: name of the vm to be catched
+        :return: the virsh session with 'event'
+        """
+        virsh_session = aexpect.ShellSession(VIRSH_EXEC)
+        event_cmd = 'event %s --all --loop' % vm_name
+        logging.info('Sending "%s" to virsh shell', event_cmd)
+        virsh_session.sendline(event_cmd)
+
+        return virsh_session
+
+    @staticmethod
+    def finish_get_event(virsh_session):
+        """
+        Stop virsh session and return the event output
+
+        :param virsh_session: virsh session to catch events
+        :return: actual event output catched
+        """
+        virsh_session.send_ctrl("^C")
+        time.sleep(5)
+        event_output = virsh_session.get_stripped_output()
+        virsh_session.close()
+        logging.debug('Event output is %s:', event_output)
+
+        return event_output
+
+    @staticmethod
+    def wait_event(func):
+        """
+        Decorator aiming to return until certain type of virsh event defined
+        by func was found
+        """
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+
+            def _get_arg_value(arg):
+                return kwargs.get(arg) if arg in kwargs else \
+                    inspect.signature(func).parameters[arg].default
+
+            def _get_event_output(session):
+                output = session.get_stripped_output()
+                logging.debug(output)
+                return output
+
+            wait_for_event = _get_arg_value('wait_for_event')
+            event_type = _get_arg_value('event_type')
+            event_timeout = _get_arg_value('event_timeout')
+
+            if wait_for_event is True and event_type is not None:
+                virsh_session = EventTracker.start_get_event(str(args[0]))
+                ret = func(*args, **kwargs)
+
+                if not utils_misc.wait_for(
+                        lambda: re.search(
+                            event_type,
+                            _get_event_output(virsh_session)
+                        ), event_timeout):
+                    raise EventNotFoundError('Not found event %s after %s seconds' %
```

Comment:
I assume we can not raise exception here to mark case failure if no events are received.
Regarding change-media, whether it received event or not is to determin how fast it will return virsh.change_media result, but not final 'pass' or 'fail' ,which should be judged on virsh.change_media result

Reply:
Sorry, I don't quite understand you. so now, you are using wait_for_event to wait for the event but you don't want to get it,right? If so, I think what you need is time.sleep.

Reply:
> Sorry, I don't quite understand you. so now, you are using wait_for_event to wait for the event but you don't want to get it,right? If so, I think what you need is time.sleep.

@Yingshun 
Let's just look at primitive implementation change_media in virsh.py
What wait_change_event() bring is to just avoid dead sleep, but not break return of change_media_result regardless whether expected events is received or not.

```
def change_media(name, device, options, wait_change_event=False, event_timeout=7, **dargs):
    """
    Change media of CD or floppy drive.

    :param name: VM's name.
    :param path: Fully-qualified path or target of disk device
    :param options: command change_media options.
    :param wait_change_event: wait until device_change event comes
    :param event_timeout: timeout for virsh event command
    :param dargs: standardized virsh function API keywords
    :return: CmdResult instance
    """
    cmd = "change-media %s %s " % (name, device)
    if options:
        cmd += " %s " % options
    change_media_result = command(cmd, **dargs)
    if wait_change_event:
        event(domain=name, event='tray-change', event_timeout=event_timeout, **dargs)

    return change_media_result

```

Reply:
@kylazhang, I assume somehow you know the  event () background

Reply:
@chunfuwen There is a very simple solution here: If you don't want to wait for an event, just set wait_for_event=False. This just perfectly meets your need.

Reply:
@kylazhang 
Maybe you can help elaborate a little bit about what original PR:https://github.com/autotest/tp-libvirt/pull/3363 want to achieve after discussion.
@hahakiki2010 
Maybe, you can help describe more about original purposes why add wait_for_event into detach_device,detach_disk,detach_device_alias,detach_interface those methods

Reply:
> @chunfuwen There is a very simple solution here: If you don't want to wait for an event, just set wait_for_event=False. This just perfectly meets your need.

@chloerh 
NO, that's not what current change_media,detach_device,detach_disk,detach_device_alias,detach_interface achieved.

Reply:
> @chunfuwen There is a very simple solution here: If you don't want to wait for an event, just set wait_for_event=False. This just perfectly meets your need.

@chloerh 
NO, that's not what current change_media,detach_device,detach_disk,detach_device_alias,detach_interface achieved.

Reply:
@chunfuwen It is. You would find that I'm right about it after reading the code. Thanks.

Reply:
I think this comment should be resolved by my last comments...